### PR TITLE
Port atomic update feature from ftw.solr into collective.solr.

### PR DESCRIPTION
**Pull request by maethu**
This PR imlements the atomic udpate feature from ftw.solr into collective.solr. 

In order to fully support the atomic update feature introduced in Solr 4 the following changes are required:
1. All fields need in your solr.cfg needs the `stored:true`attribute
2. `<updateLog/>` configured in solrconfig.xml (Supported by collective.receipe.solrinstance 6.0.0b3)
3. A new field:  `_version_`

More details --> http://wiki.apache.org/solr/Atomic_Updates

Example:
Call `solr-maintenance/reindex?idxs:list=Title&idxs:list=Description` on any object.

```
2015/09/15-15:39:12 reindexing solr catalog...
2015/09/15-15:39:12 indexing only ['Title', 'Description'] 
2015/09/15-15:39:12 indexing <Folder at /Plone/downloads>
2015/09/15-15:39:12 indexing <File at /Plone/downloads/file5>
2015/09/15-15:39:12 indexing <File at /Plone/downloads/file4>
2015/09/15-15:39:12 indexing <File at /Plone/downloads/file3>
2015/09/15-15:39:12 indexing <File at /Plone/downloads/file2>
2015/09/15-15:39:12 indexing <File at /Plone/downloads/file1>
2015/09/15-15:39:12 intermediate commit (6 items processed, last batch in 0.024s)...
2015/09/15-15:39:12 solr index rebuilt.
2015/09/15-15:39:12 processed 6 items in 0.055s (0.027s cpu time).
```

### CODE REVIEW

**Comment by tisto** [comment]
@maethu Nice! Travis tests seem to fail though.

File: docs/CHANGES.rst [hunk 1]
```diff
@@ -4,6 +4,19 @@ Changelog
 4.1.1 (unreleased)
 ------------------
 
+- Ported atomic updates from ftw.solr.
```
Comment (tisto) [inline docs/CHANGES.rst]: @maethu wouldn't it make sense to add this to our new sphinx-based docs folder and just link to that here?

File: solr.cfg [hunk 2]
```diff
@@ -39,12 +39,13 @@ java_opts =
   -server
   -Xms${settings:solr-min-ram}
   -Xmx${settings:solr-max-ram}
+
 index =
-    name:allowedRolesAndUsers   type:string stored:false multivalued:true
```
Comment (tisto) [inline solr.cfg]: @maethu could you elaborate why we have to store those values? Why would we want to return the "default" or the "allowedRolesAndUsers" field in our search results? Is this even related to the new feature?

**Comment by maethu** [comment]
@tisto Yes I'm aware of that, but currently my local build of collective.solr does not work out of the box :disappointed:

File: solr.cfg (hunk 2 above)
Comment (maethu) [inline solr.cfg]: @tisto 
According to the wiki page on http://wiki.apache.org/solr/Atomic_Updates:

>The core functionality of atomically updating a document requires that all fields in your SchemaXml must be configured as stored="true" except for fields which are <copyField/> destinations -- which must be configured as stored="false". This is because the atomic updates are applied to the document represented by the existing stored field values.

So this a solr requirement.

**Comment by tisto** [comment]
@maethu what's the problem with your local build? Do you get a traceback?

File: solr.cfg (hunk 2 above)
Comment (tisto) [inline solr.cfg]: ic. The "default" field should be "stored=false" then though, according to the Solr docs.

**Comment by maethu** [comment]
@tisto  yep here https://github.com/collective/collective.solr/issues/79
With some small changes, it's now up and running :smile:

**Comment by tisto** [comment]
@maethu travis tests are still failing though.

**Comment by maethu** [comment]
@tisto I'm still working on it. On my local machine I got still 3 failing tests.

**Comment by tisto** [comment]
Cool. Though, I would recommend to separate the p.r.solrinstance upgrade from your current work. Otherwise it will be hard to figure out why the test output started to change (at least for the pr reviewer). 

We have to take into account that we want to test against different Solr/Plone versions and I'm still somehow sceptical about the c.r.solrinstance upgrade. As far as I know version 6.0.x was a complete rewrite of the recipe and last time I looked into it I wasn't really ready. I would feel better if we would stick with the stable 5.x branch for now.

**Comment by maethu** [comment]
@tisto 
OK, I'll move the c.r.solrinstance 6.x changes into it's own branch. 

Unfortunately I'm still struggling with ONE failing test...
I'm no able to get the boost value feature back.

**Comment by maethu** [comment]
@tisto OK I moved the c.r.solrinstance 6 feature into a new branch --> https://github.com/collective/collective.solr/tree/mle-collective-receipe-solrinstance-6

**Comment by tisto** [comment]
@maethu awesome! thanks!

**Comment by maethu** [comment]
@tisto This is somehow frustrating... with c.r.solrinstance 5 I'm unable to run the buildout. I changed the method `is_solr_4` in the receipe to always return `True`, but this somehow destroys the solrconfig.xml, so it does no longer recognise the `updateLog`. You may check build on travis. 

The solr.log says:
`ERROR org.apache.solr.core.SolrCore  – org.apache.solr.common.SolrException: Atomic document updates are not supported unless <updateLog/> is configured`
But imho it is. I'm totally stuck with this and need help. 

I now go ahead with c.r.solrinstance 6.

**Comment by tisto** [comment]
@maethu ok, I understand. Then go ahead with c.r.solrinstance 6. It's great that you are working on this and I would really like to have atomic updates feature merged.

**Comment by maethu** [comment]
Closed in favour of https://github.com/collective/collective.solr/pull/82
